components/binding-usage-controller/internal/controller/usagekind/protection_controller_test.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/kyma-project/kyma/components/binding-usage-controller/pkg/client/clientset/versioned/fake"
+	informers "github.com/kyma-project/kyma/components/binding-usage-controller/pkg/client/informers/externalversions"
+)
+
+func TestProtectionControllerAddsFinalizer(t *testing.T) {
+	// GIVEN
+	tc := newProtectionControllerTestCase()
+	defer tc.TearDown()
+	tc.RunController()
+
+	// WHEN
+	tc.AddUsageKind(t, fixUsageKind())
+	tc.WaitForAddFinalizer(t, 3*time.Second)
+
+	// THEN
+	uk := tc.GetUsageKind(t, fixUsageKind().Name)
+	assert.Contains(t, uk.Finalizers, usagekind.FinalizerName)
+
+	// WHEN
+	tc.MarkRemoval(t, uk.Name)
+	tc.WaitForProcessDeletion(t, 3*time.Second)
+
+	// THEN
+	uk = tc.GetUsageKind(t, fixUsageKind().Name)
+	assert.NotContains(t, uk.Finalizers, usagekind.FinalizerName)
+}
+
+func TestProtectionControllerProtectsUsageKindUsedBySBU(t *testing.T) {
+	// GIVEN
+	tc := newProtectionControllerTestCase()
+	defer tc.TearDown()
+	tc.RunController()
+
+	tc.AddSBU(t, fixSBU(fixUsageKind().Name))
+	tc.AddUsageKind(t, fixUsageKind())
+	tc.WaitForAddFinalizer(t, 3*time.Second)
+
+	// WHEN
+	tc.MarkRemoval(t, fixUsageKind().Name)
+	tc.WaitForProcessDeletion(t, 3*time.Second)
+
+	// THEN
+	uk := tc.GetUsageKind(t, fixUsageKind().Name)
+	assert.Contains(t, uk.Finalizers, usagekind.FinalizerName)
+
+	// WHEN
+	tc.RemoveSBU(t, fixSBU(fixUsageKind().Name))
+	tc.WaitForProcessDeletion(t, 3*time.Second)
+
+	// THEN
+	uk = tc.GetUsageKind(t, fixUsageKind().Name)
+	assert.NotContains(t, uk.Finalizers, usagekind.FinalizerName)
+}
+
+func fixSBU(kind string) *v1alpha1.ServiceBindingUsage {
+	return &v1alpha1.ServiceBindingUsage{
+		ObjectMeta: metaV1.ObjectMeta{
+			Namespace: "production",
+			Name:      "sbu-",
+			UID:       "uid-123",
+		},
+		Spec: v1alpha1.ServiceBindingUsageSpec{
+			UsedBy: v1alpha1.LocalReferenceByKindAndName{
+				Name: "redis-client",
+				Kind: kind,
+			},
+			ServiceBindingRef: v1alpha1.LocalReferenceByName{
+				Name: "redis-client",
+			},
+		},
+	}
+}
+
+type protectionTestCase struct {
+	*testCase
+
+	addFinalizerOpDone    chan struct{}
+	processDeletionOpDone chan struct{}
+}
+
+func newProtectionControllerTestCase() *protectionTestCase {
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+
+	addFinalizerDone := make(chan struct{})
+	addFinalizerOpHook := func() {
+		addFinalizerDone <- struct{}{}
+	}
+
+	processDeletionDone := make(chan struct{})
+	processDeletionOpHook := func() {
+		processDeletionDone <- struct{}{}
+	}
+
+	cs := fake.NewSimpleClientset()
+	informersFactory := informers.NewSharedInformerFactory(cs, 0)
+
+	ctrl := usagekind.NewProtectionController(informersFactory.Servicecatalog().V1alpha1().UsageKinds(),
+		informersFactory.Servicecatalog().V1alpha1().ServiceBindingUsages(),
+		cs.ServicecatalogV1alpha1(),
+		spy.NewLogSink().Logger).
+		WithTestHookOnAsyncOpDone(addFinalizerOpHook, processDeletionOpHook)
+
+	informersFactory.Start(ctx.Done())
```

Comment:
and what about `informersFactory.WaitForCacheSync`??
test can be not deterministic without that

Reply:
testCase.RunCOntroller runs the controller (method Run) -> there is a WaitForCacheSync in method Run in t the ProtectionController implementation